src/connectors/impls/gpubsub/gsub.rs

```diff
@@ -0,0 +1,279 @@
+
+                    // If we can't send to the main task, disconnect and let it restart
+                    return;
+                }
+            }
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Source for GSubSource {
+    async fn connect(&mut self, _ctx: &SourceContext, _attempt: &Attempt) -> Result<bool> {
+        let token = Token::new()?;
+
+        let tls_config = ClientTlsConfig::new()
+            .ca_certificate(Certificate::from_pem(googapis::CERTIFICATES))
+            .domain_name("pubsub.googleapis.com");
+
+        let channel = Channel::from_static("https://pubsub.googleapis.com")
+            .connect_timeout(Duration::from_nanos(self.config.connect_timeout))
+            .tls_config(tls_config)?
+            .connect()
+            .await;
+
+        let channel = dbg!(channel)?;
+
+        let client = SubscriberClient::with_interceptor(
+            channel,
+            AuthInterceptor {
+                token: Box::new(move || match token.header_value() {
+                    Ok(val) => Ok(val),
+                    Err(_) => Err(Status::unavailable(
+                        "Failed to retrieve authentication token.",
+                    )),
+                }),
+            },
+        );
+
+        let client = Arc::new(Mutex::new(client));
+
+        let (tx, rx) = async_std::channel::bounded(QSIZE.load(Ordering::Relaxed));
+
+        async_std::task::spawn(consumer_task(
+            client.clone(),
+            tx,
+            self.ack_ids.clone(),
+            self.config.subscription_id.clone(),
+            Duration::from_secs(10), // fixme get this from the config
+        ));
+
+        self.receiver = Some(rx);
+        self.client = Some(client);
+
+        Ok(true)
+    }
+
+    async fn pull_data(&mut self, pull_id: &mut u64, ctx: &SourceContext) -> Result<SourceReply> {
+        let receiver = self
+            .receiver
+            .as_mut()
+            .ok_or(ErrorKind::BigQueryClientNotAvailable(
+                // fixme use an error for GPubSub
+                "The receiver is not connected",
+            ))?;
+
+        let (ack_id, data) = match receiver.recv().await? {
+            Ok(response) => response,
+            Err(error) => {
+                let error_kind = error.kind();
+                return if let ErrorKind::Timeout(_) = error_kind {
+                    ctx.swallow_err(
+                        ctx.notifier.connection_lost().await,
+                        "Failed to notify about PubSub connection loss",
+                    );
+
+                    // fixme use the actual stream id
+                    Ok(SourceReply::StreamFail(0))
+                } else {
+                    Err(error)
+                };
+            }
+        };
+
+        *pull_id = ack_id;
+
+        Ok(SourceReply::Data {
+            origin_uri: EventOriginUri::default(),
+            data,
+            meta: None,
+            stream: None,
```

Comment:
If gsub has the notion of a partition, we should somehow use 1 stream per partition, as we do in kafka. The reason is that we only have order inside of single streams.

Reply:
There's no partitions in gpubsub, messages can have an ordering key, but then the ordering is handled internally by the platform (https://cloud.google.com/pubsub/docs/ordering), i.e. the messages received by sending PullRequest will always be in order. I don't think there's an easy way to split it into separate streams, since there's no pre-defined set of ordering keys, they can come and go.

Reply:
then lets just use `Some(DEFAULT_STREAM_ID)` for the time being